besskge/scoring.py

```diff
@@ -602,6 +638,254 @@ def score_tails(
         relation_emb = torch.index_select(
             self.relation_embedding, index=relation_id, dim=0
         )
-        return self.broadcasted_score(
+        return self.broadcasted_dot_product(
             complex_multiplication(head_emb, relation_emb), tail_emb
         )
+
+
+class BoxE(DistanceBasedScoreFunction):
+    """
+    BoxE scoring function :cite:p:`BoxE`.
+    """
+
+    def __init__(
+        self,
+        negative_sample_sharing: bool,
+        scoring_norm: int,
+        sharding: Sharding,
+        n_relation_type: int,
+        embedding_size: int,
+        entity_initializer: Union[torch.Tensor, List[Callable[..., torch.Tensor]]] = [
+            torch.nn.init.uniform_,
+        ],
+        relation_initializer: Union[torch.Tensor, List[Callable[..., torch.Tensor]]] = [
+            torch.nn.init.uniform_,
+            init_uniform_norm,
+        ],
+        apply_tanh: bool = True,
+        dist_func_per_dim: bool = True,
+    ) -> None:
+        """
+        Initialize BoxE model.
+
+        :param negative_sample_sharing:
+            see :meth:`DistanceBasedScoreFunction.__init__`
+        :param scoring_norm:
+            see :meth:`DistanceBasedScoreFunction.__init__`
+        :param sharding:
+            Entity sharding.
+        :param n_relation_type:
+            Number of relation types in the knowledge graph.
+        :param embedding_size:
+            Size of final entity embeddings.
+        :param entity_initializer:
+            Initialization function or table for entity embeddings.
+        :param relation_initializer:
+            Initialization functions or table for relation embeddings.
+            If not passing a table, two functions are needed: the initializer
+            for head/tail box centers and the initializer for
+            (scalar) head/tail box sizes.
+        :param apply_tanh:
+            If True, bound relation box sizes and bumped entity
+            representations with tanh.
+        :param dist_func_per_dim:
+            If True, instead of selecting between the two BoxE distance
+            functions based on whether the bumped representation is inside
+            or outside the relation box, make the choice separately for
+            each dimension of the embedding space.
+        """
+        super(BoxE, self).__init__(
+            negative_sample_sharing=negative_sample_sharing, scoring_norm=scoring_norm
+        )
+        self.apply_tanh = apply_tanh
+        self.dist_func_per_dim = dist_func_per_dim
+        self.soft = 1e-6
```

Comment:
Should we have this as a param that defaults to 1e-6? And maybe call it `eps` or similar - would be more intuitive to me :)